src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalSinglePhasePoromechanics_impl.hpp

```diff
@@ -0,0 +1,491 @@
+  real64 const detJxW = m_finiteElementSpace.template getGradN< FE_TYPE >( k, q, stack.xLocal,
+                                                                           stack.feStack, dNdX );
+
+  // Step 2: compute strain increment
+  real64 strainIncrement[6]{};
+  FE_TYPE::symmetricGradient( dNdX, stack.uhat_local, strainIncrement );
+
+  // Step 3: compute 1) the total stress, 2) the body force terms, and 3) the fluidMassIncrement
+  // using quantities returned by the PorousSolid constitutive model.
+  // This function also computes the derivatives of these three quantities wrt primary variables
+  smallStrainUpdate( k, q, strainIncrement, stack );
+
+  // Step 4: use the total stress and the body force to increment the local momentum balance residual
+  // This function also fills the local Jacobian rows corresponding to the momentum balance.
+  assembleMomentumBalanceTerms( N, dNdX, detJxW, stack );
+
+  // Step 5: use the fluid mass increment to increment the local mass balance residual
+  // This function also fills the local Jacobian rows corresponding to the mass balance.
+  assembleElementBasedFlowTerms( dNdX, detJxW, stack );
+}
+
+template< typename SUBREGION_TYPE,
+          typename CONSTITUTIVE_TYPE,
+          typename FE_TYPE >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+real64 ThermalSinglePhasePoromechanics< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >::
+complete( localIndex const k,
+          StackVariables & stack ) const
+{
+  real64 const maxForce = Base::complete( k, stack );
+
+  localIndex const numSupportPoints =
+    m_finiteElementSpace.template numSupportPoints< FE_TYPE >( stack.feStack );
+  integer numDisplacementDofs = numSupportPoints * numDofPerTestSupportPoint;
+
+  // Step 1: assemble the derivatives of linear momentum balance wrt temperature into the global matrix
+
+  for( int localNode = 0; localNode < numSupportPoints; ++localNode )
+  {
+    for( integer dim = 0; dim < numDofPerTestSupportPoint; ++dim )
+    {
+      localIndex const dof = LvArray::integerConversion< localIndex >( stack.localRowDofIndex[numDofPerTestSupportPoint*localNode + dim] - m_dofRankOffset );
+      if( dof < 0 || dof >= m_matrix.numRows() )
+      {
+        continue;
+      }
+
+      m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                              &stack.localTemperatureDofIndex,
+                                                                              stack.dLocalResidualMomentum_dTemperature[numDofPerTestSupportPoint * localNode + dim],
+                                                                              1 );
+    }
+  }
+
+  // Step 2: assemble the derivatives of mass balance residual wrt temperature into the global matrix
+
+  localIndex const massDof = LvArray::integerConversion< localIndex >( stack.localPressureDofIndex - m_dofRankOffset );
+  if( 0 <= massDof && massDof < m_matrix.numRows() )
```

Comment:
Same?

Reply:
Yes, this is the same here, I added a comment to explain that we need to filter out ghost cells in the assembly of the Jacobian matrix, and then I propagated the comment.